File: frontend/src/api/runtime.ts
export type RuntimeOverallStatus = "ready" | "running" | "attention" | "blocked" | "unknown";
export type RuntimeServiceStatus = "ready" | "starting" | "stopping" | "stopped" | "unreachable" | "missing" | "unhealthy" | "unknown";
export type RuntimeOperationStatus = "queued" | "running" | "succeeded" | "failed" | "blocked" | "timed_out" | "cancelled" | "interrupted";

export interface RuntimeProbeStatus {
  name: string;
  status: RuntimeServiceStatus;
  detail: string;
  url: string | null;
}

export interface RuntimePortStatus {
  name: string;
  host: string;
  port: number;
  status: RuntimeServiceStatus;
  detail: string;
}

export interface RuntimeContainerStatus {
  name: string;
  required: boolean;
  exists: boolean;
  running: boolean;
  status: RuntimeServiceStatus;
  statusText: string | null;
}

export interface RuntimeConfigItem {
  key: string;
  label: string;
  value: string;
  source: string;
  secret: boolean;
}

export interface RuntimeOperation {
  operationId: string;
  kind: "start" | "stop";
  status: RuntimeOperationStatus;
  requestedAt: string;
  startedAt: string | null;
  finishedAt: string | null;
  actor: string;
  errorCode: string | null;
  errorMessage: string | null;
}

export interface RuntimeStatusResponse {
  overallStatus: RuntimeOverallStatus;
  reasonCode: string;
  reasonText: string;
  checkedAt: string;
  projectPath: string;
  projectId: string;
  docker: RuntimeProbeStatus;
  wsl: RuntimeProbeStatus;
  supabaseCli: RuntimeProbeStatus;
  api: RuntimePortStatus;
  db: RuntimePortStatus;
  studio: RuntimePortStatus;
  edgeRuntime: RuntimeProbeStatus;
  grafana: RuntimeProbeStatus;
  containers: RuntimeContainerStatus[];
  config: RuntimeConfigItem[];
  activeOperation: RuntimeOperation | null;
}

export interface RuntimeOperationCreateResponse {
  operationId: string;
  status: RuntimeOperationStatus;
  detailUrl: string;
}

export async function fetchLocalSupabaseStatus(): Promise<RuntimeStatusResponse> {
  const response = await fetch("/api/runtime/local-supabase");
  if (!response.ok) throw new Error("Local Supabase status could not be loaded");
  return response.json();
}

export async function startLocalSupabase(): Promise<RuntimeOperationCreateResponse> {
  return mutateRuntime("start");
}

export async function stopLocalSupabase(): Promise<RuntimeOperationCreateResponse> {
  return mutateRuntime("stop");
}

async function mutateRuntime(action: "start" | "stop"): Promise<RuntimeOperationCreateResponse> {
  const response = await fetch(`/api/runtime/local-supabase/${action}`, { method: "POST" });
  if (!response.ok) {
    const raw = await response.json().catch(() => null);
    throw new Error(raw?.detail?.reason ?? `Local Supabase ${action} failed`);
  }
  return response.json();
}

File: backend/app/schemas/runtime.py
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import Field

from backend.app.schemas.state_context import StateContextDto, unknown_state_context
from backend.app.schemas.upload_preview import ApiModel


class RuntimeOverallStatus(str, Enum):
    ready = "ready"
    running = "running"
    attention = "attention"
    blocked = "blocked"
    unknown = "unknown"


class RuntimeServiceStatus(str, Enum):
    ready = "ready"
    starting = "starting"
    stopping = "stopping"
    stopped = "stopped"
    unreachable = "unreachable"
    missing = "missing"
    unhealthy = "unhealthy"
    unknown = "unknown"


class RuntimeOperationKind(str, Enum):
    start = "start"
    stop = "stop"


class RuntimeOperationStatus(str, Enum):
    queued = "queued"
    running = "running"
    succeeded = "succeeded"
    failed = "failed"
    blocked = "blocked"
    timed_out = "timed_out"
    cancelled = "cancelled"
    interrupted = "interrupted"


class RuntimeEventLevel(str, Enum):
    info = "info"
    warning = "warning"
    error = "error"


class RuntimeContainerStatus(ApiModel):
    name: str
    required: bool
    exists: bool
    running: bool
    status: RuntimeServiceStatus
    status_text: str | None = None


class RuntimePortStatus(ApiModel):
    name: str
    host: str = "127.0.0.1"
    port: int
    status: RuntimeServiceStatus
    detail: str


class RuntimeProbeStatus(ApiModel):
    name: str
    status: RuntimeServiceStatus
    detail: str
    url: str | None = None


class RuntimeConfigItem(ApiModel):
    key: str
    label: str
    value: str
    source: str
    secret: bool = False


class RuntimeOperationDto(ApiModel):
    operation_id: str
    kind: RuntimeOperationKind
    status: RuntimeOperationStatus
    requested_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    actor: str
    error_code: str | None = None
    error_message: str | None = None


class RuntimeEventDto(ApiModel):
    event_id: int
    operation_id: str
    seq: int
    ts: datetime
    level: RuntimeEventLevel
    event_type: str
    message: str
    data: dict[str, Any] = Field(default_factory=dict)


class RuntimeStatusResponse(ApiModel):
    overall_status: RuntimeOverallStatus
    reason_code: str
    reason_text: str
    checked_at: datetime
    project_path: str
    project_id: str
    docker: RuntimeProbeStatus
    wsl: RuntimeProbeStatus
    supabase_cli: RuntimeProbeStatus
    api: RuntimePortStatus
    db: RuntimePortStatus
    studio: RuntimePortStatus
    edge_runtime: RuntimeProbeStatus
    grafana: RuntimeProbeStatus
    containers: list[RuntimeContainerStatus]
    config: list[RuntimeConfigItem]
    state_context: StateContextDto = Field(default_factory=unknown_state_context)
    active_operation: RuntimeOperationDto | None = None


class RuntimeOperationCreateResponse(ApiModel):
    operation_id: str
    status: RuntimeOperationStatus
    detail_url: str


class RuntimeOperationDetailResponse(ApiModel):
    operation: RuntimeOperationDto
    events: list[RuntimeEventDto]
